### ISSUE TITLE
In nested decorator, a reassigned variable's type is not tracked

### ISSUE THREAD

**Issue by sirosen** [python/mypy#9720]
**Bug Report**

Apologies if this is a duplicate. I could not find this reported, but it might have appeared with very different wording.

Reassigning a variable typed as Union detects the new type of the variable.
However, if the variable's usage site is in a function, the original (not the detected) type is used, even if the variable is never reassigned.

This comes up naturally in the context of a function which returns a decorator and which reassigns one of its arguments to make it conform to the requirements of the decorator.

**To Reproduce**

I ran into this working with `marshmallow` Schema objects, so I'll use that as my example.
Running mypy on the following file `repro.py` fails:

```python
import functools
import typing

from marshmallow import Schema


def apply_dump_schema(
    output: typing.Union[Schema, typing.Type[Schema]] = None
) -> typing.Callable:
    if output and not isinstance(output, Schema):
        output = output()

    def decorator(fn):
        @functools.wraps(fn)
        def wrapped(*args, **kwargs):
            data = fn(*args, **kwargs)
            if output:
                data = output.dump(data)
            return data

        return wrapped

    return decorator
```

**Expected Behavior**

`mypy` should pass.

It should see the first `if` and deduce that `output`'s type is now `Optional[Schema]` (i.e. `Type[Schema]` is no longer possible).
It should then see the `if output:` clause and deduce that the type is `Schema`, so `output.dump(data)` is okay.

**Actual Behavior**

`mypy` appears to determine that the type of `output` is `Type[Schema]` -- that it is a Schema class and not a Schema instance.

The error is:
```
repro.py:18: error: Too few arguments for "dump" of "Schema"
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 0.782, 0.790
- Python version used: 3.6

**Additional Note**

A very similar case works just fine, which is surprising for a naive user:
```python
import functools
import typing

from marshmallow import Schema


def apply_dump_schema(
    output: typing.Union[Schema, typing.Type[Schema]] = None
) -> typing.Callable:
    if isinstance(output, type):
        output_s: typing.Optional[Schema] = output()
    else:
        output_s = output

    def decorator(fn):
        @functools.wraps(fn)
        def wrapped(*args, **kwargs):
            data = fn(*args, **kwargs)
            if output_s:
                data = output_s.dump(data)
            return data

        return wrapped

    return decorator
```

In this case, `output: Optional[Union[Schema, Type[Schema]]]` is replaced with `output_s: Optional[Schema]`.
I believe that what makes this work is the more restrictive type for `output_s`, so mypy can be certain that `output_s` is never later reassigned to be a `Type[Schema]`. But this should work in the original case as well because we can see that `output` is never reassigned.

Is it possible for `mypy` to recognize in the original case what the type of `output` should be?

**Comment by hauntsaninja** [comment]
Duplicate of #2608, #9197, #9325 and others

**Comment by sirosen** [comment]
Ah, really sorry for my deficient search of open issues. Thanks for triaging/redirecting! I'll watch #2608 now.
